plugins/parsers/opentsdb/parser.go

```diff
@@ -0,0 +1,127 @@
+package opentsdb
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+// Parser encapsulates a OpenTSDB Parser.
+type Parser struct {
+	DefaultTags map[string]string `toml:"-"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	var metrics []telegraf.Metric
+
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		// delete LF and CR
+		if line[len(line)-1] == '\n' {
+			line = line[:len(line)-1]
+		}
+		if line[len(line)-1] == '\r' {
+			line = line[:len(line)-1]
+		}
+
+		m, err := p.ParseLine(line)
+		if err != nil {
+			return nil, err
+		}
+
+		if m != nil {
+			metrics = append(metrics, m)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return metrics, nil
+}
+
+// ParseLine performs OpenTSDB parsing of a single line.
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	// Break into fields ("put", name, timestamp, value, tag1, tag2, ..., tagN).
+	fields := strings.Fields(line)
+	if len(fields) == 0 {
+		return nil, nil
+	}
+	if len(fields) < 4 || fields[0] != "put" {
+		return nil, fmt.Errorf("received %q which doesn't have required fields", line)
```

Comment:
Should we really error out here  (especially if the first item is _not_ "put") or would just ignoring the metric would also be good?

Reply:
I'm inclined to reject the whole if any part of it is invalid, but I don't have a super-strong feeling. Happy to go with what you prefer here - just let me know.

Reply:
My question is more, if the protocol is extended to contain other items than `put` you get an error and that error then leads to all other metrics being dropped in `Parse` after the first invalid line. This might be somewhat unexpected for the user. ;-)
But thinking about this more, how about error-out here but _log and continue_ in `Parse` if `ParseLine` throws an error?

Reply:
OK, that seems reasonable to me :thumbsup: . I've implemented this in e36d3a8c1 ; what do you think?

I copied the logger implementation from the `grok` plugin, and it does work in my whole-process test, but not explicitly initializing the `Log` field anywhere outside of tests makes me nervous :sweat_smile:. I assume there's some general code somewhere that does it, but wasn't able to find it with a quick search.

Reply:
The agent does this for you. The code is [here](https://github.com/influxdata/telegraf/blob/a57916843768f9b8bbf56b356e0c9b8cf0f22085/agent/agent.go#L211).